src/main/java/org/reactome/addlinks/referencecreators/EnsemblBioMartMicroarrayPopulator.java

```diff
@@ -0,0 +1,159 @@
+                // Iterate through each identifier, and then through each RGP instance.
+                for (String rgpIdentifier : rgpIdentifiersToRGPs.keySet()) {
+                    for (GKInstance rgpInst : rgpIdentifiersToRGPs.get(rgpIdentifier)) {
+                        addMicroarrayDataToReferenceGeneProduct(rgpIdentifier, rgpInst, speciesBiomartName, mappings);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Iterates through each protein associated with the RGP instance and finds transcripts associated with protein identifier.
+     * Then takes transcript identifier and finds all microarray identifiers associated with it. These microarray identifiers
+     * are then added to the 'otherIdentifier' attribute of the RGP instance. Once all microarray identifiers have been added,
+     * the instance is updated with the new data in the DB.
+     * @param rgpIdentifier - String, Identifier associated with RGP instance.
+     * @param rgpInst - GKInstance, ReferenceGeneProduct instance
+     * @param speciesBiomartName, - String, Species name in BioMart format (eg: hsapiens).
+     * @param mappings - Map<String, Map< String, List<String>>>, Super mapping structure generated by EnsemblBioMartFileProcessor.
+     * @throws Exception - Thrown by MySQLAdaptor/GKInstance methods. Typically due to accessing data that doesn't exist or from InvalidAttributeExceptions.
+     */
+    private void addMicroarrayDataToReferenceGeneProduct(String rgpIdentifier, GKInstance rgpInst, String speciesBiomartName, Map<String, Map<String, List<String>>> mappings) throws Exception {
+        // Retrieve protein identifier associated with reference DB.
+        Set<String> rgpProteins = findRGPProteins(rgpIdentifier, rgpInst, speciesBiomartName, mappings);
+        // For each protein identifier retrieved, find microarray identifiers associated and
+        // add them to the 'otherIdentifier' attribute of the RGP instance.
+        for (String protein : rgpProteins) {
+            if (EnsemblBioMartUtil.getProteinToTranscriptsMappings(speciesBiomartName, mappings).containsKey(protein)) {
+                for (String transcript : EnsemblBioMartUtil.getProteinToTranscriptsMappings(speciesBiomartName, mappings).get(protein)) {
+                    if (EnsemblBioMartUtil.getTranscriptToMicroarraysMappings(speciesBiomartName, mappings).containsKey(transcript)) {
+                        for (String microarrayId : EnsemblBioMartUtil.getTranscriptToMicroarraysMappings(speciesBiomartName, mappings).get(transcript)) {
+                            Collection<String> otherIdentifiers = rgpInst.getAttributeValuesList(ReactomeJavaConstants.otherIdentifier);
+                            if (!otherIdentifiers.contains(microarrayId) && !this.testMode) {
+                                rgpInst.addAttributeValue(ReactomeJavaConstants.otherIdentifier, microarrayId);
+                            }
+                        }
```

Comment:
Currently, with each new microarray identifier, the database is being accessed and added to.  It might be more efficient first to retrieve the `otherIdentifiers` from the ReferenceGeneProduct then loop over and add all new microarray ids (they can be stored in set so duplicates will get dealt with automatically when adding new values).  Once `otherIdentifiers` is updated, it can replace the existing values in the otherIdentifier attribute of the ReferenceGeneProduct:

```
Set<String> otherIdentifiers = new HashSet<String>(rgpInst.getAttributeValuesList(ReactomeJavaConstants.otherIdentifier));
for (String microarrayId : getMicroarrayIdentifiers(proteinIdentifier)) {
    otherIdentifiers.add(microarrayId);
}

if (!EnsemblBioMartMicroarrayPopulator.this.testMode) {
    rgpInst.setAttributeValue(ReactomeJavaConstants.otherIdentifier, otherIdentifiers);
}
```

Reply:
Something about this approach seems to be adding _a lot_ of runtime

Reply:
Yea, in 15 minutes it only added 12000 otherIdentifiers to the DB, but the original way takes only 3-4 minutes to do everything

Reply:
Oh, interesting!  Maybe overwriting the value with "setAttributeValue" takes longer than appending with "addAttributeValue".  If it's making things worse, then feel free to disregard this suggestion!